pseudoanonymize/base.py

```diff
@@ -0,0 +1,73 @@
+import logging
+import os
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional
+
+from faker import Faker
+
+from .exceptions import MaxRetriesExceededException, UnparsableLLMOutputException
+
+fake = Faker()
+
+
+class BaseProcessor(ABC):
+    def __init__(self, client: Any):
+        self.client = client
+        self.logger = self._setup_logger()
+
+    @abstractmethod
+    def predict(self, input_: Dict[str, Any]) -> Dict[str, Any]:
+        pass
+
+    def retry_prediction(self, input_: Dict[str, Any], max_retries: int = 5) -> Optional[Dict[str, Any]]:
+        retries = 0
+        while retries < max_retries:
+            try:
+                prediction = self.predict(input_)
+                return prediction
+            except UnparsableLLMOutputException as e:
+                retries += 1
+                if retries >= max_retries:
+                    self.logger.error(f"Failed to process item after {max_retries} attempts")
+                    raise MaxRetriesExceededException(f"Failed to process item after {max_retries} attempts")
+
+    def _setup_logger(self) -> logging.Logger:
+        logger = logging.getLogger(self.__class__.__name__)
+        logger.setLevel(logging.DEBUG)
+        log_dir = 'logs'
+        os.makedirs(log_dir, exist_ok=True)
+        handler = logging.FileHandler(os.path.join(log_dir, f'{self.__class__.__name__}.log'))
+        handler.setLevel(logging.DEBUG)
+        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+        handler.setFormatter(formatter)
+        logger.addHandler(handler)
+        return logger
+
+    def _log_error(self, error_message: str, combined_output: str):
+        self.logger.error(f"Error: {error_message}\nCombined Output: {combined_output}")
+
+    def _parse_replacements(self, text, replacement_dict):
```

Comment:
This method and the one below could use a docstring to explain what they do - not obvious from the name or params.

Also add missing type hints

Reply:
Done, also improved _parse_replacemtns by calling the flatten_dict function in utils.py.